src/main/java/htsjdk/samtools/BAMIndexMerger.java

```diff
@@ -0,0 +1,254 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2018 The Broad Institute
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package htsjdk.samtools;
+
+import htsjdk.samtools.seekablestream.SeekableStream;
+import htsjdk.samtools.util.BlockCompressedFilePointerUtil;
+
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Merges BAM index files for (headerless) parts of a BAM file into a single
+ * index file. The index files must have been produced using an uninitialized window (TODO).
+ */
+public class BAMIndexMerger {
+
+    private static final int UNINITIALIZED_WINDOW = -1;
+
+    /**
+     * Merge BAI files for (headerless) BAM file parts.
+     *
+     * @param header      the header for the file
+     * @param partLengths the lengths, in bytes, of the headerless BAM file parts
+     * @param baiStreams  streams for the BAI files to merge
+     * @param baiOut      the output stream for the resulting merged BAI
+     */
+    public static void merge(
+            final SAMFileHeader header,
+            final List<Long> partLengths,
+            final List<SeekableStream> baiStreams,
+            final OutputStream baiOut) {
+        if (baiStreams.isEmpty()) {
+            throw new IllegalArgumentException("Cannot merge zero BAI files");
+        }
+        final SAMSequenceDictionary dict = header.getSequenceDictionary();
+        final List<AbstractBAMFileIndex> bais = new ArrayList<>();
+        for (SeekableStream baiStream : baiStreams) {
```

Comment:
The index files are all read into memory (in the test) so each call to `AbstractBAMFileIndex` is a a memory-backed index, so there's probably not much call for a parallel stream. It would be easy to change later if it showed a performance improvement.